runtime/sema/entitlementset.go

```diff
@@ -0,0 +1,191 @@
+}
+
+// Merge merges the other entitlement set into this set.
+// The result is the union of the entitlements and disjunctions of both sets.
+//
+// The result is not necessarily minimal:
+// For example, if s contains a disjunction d,
+// and other contains an entitlement e that is part of d,
+// then the result will still contain d.
+// See Add.
+// Call Minimize to obtain a minimal set.
+func (s *EntitlementSet) Merge(other *EntitlementSet) {
+	if other.Entitlements != nil {
+		other.Entitlements.Foreach(func(key *EntitlementType, _ struct{}) {
+			s.Add(key)
+		})
+	}
+
+	if other.Disjunctions != nil {
+		other.Disjunctions.
+			Foreach(func(_ string, disjunction *EntitlementOrderedSet) {
+				s.AddDisjunction(disjunction)
+			})
+	}
+}
+
+// Minimize minimizes the entitlement set.
+// It removes disjunctions that contain entitlements
+// which are also in the entitlement set
+func (s *EntitlementSet) Minimize() {
+	// If there are no entitlements or no disjunctions,
+	// there is nothing to minimize
+	if s.Entitlements == nil || s.Disjunctions == nil {
+		return
+	}
+
+	// Remove disjunctions that contain entitlements that are also in the entitlement set
+	var keysToRemove []string
+	s.Disjunctions.Foreach(func(key string, disjunction *EntitlementOrderedSet) {
+		if disjunction.ForAnyKey(s.Entitlements.Contains) {
+			keysToRemove = append(keysToRemove, key)
+		}
+	})
+
+	for _, key := range keysToRemove {
+		s.Disjunctions.Delete(key)
+	}
+}
+
+// Access returns the access represented by the entitlement set.
+// The set is minimized before the access is computed.
+func (s *EntitlementSet) Access() Access {
+	if s == nil {
+		return UnauthorizedAccess
+	}
+
+	s.Minimize()
+
+	var entitlements *EntitlementOrderedSet
+	if s.Entitlements != nil && s.Entitlements.Len() > 0 {
+		entitlements = orderedmap.New[EntitlementOrderedSet](s.Entitlements.Len())
+		entitlements.SetAll(s.Entitlements)
+	}
+
+	if s.Disjunctions != nil && s.Disjunctions.Len() > 0 {
+		if entitlements == nil {
+			// If there are no entitlements, and there is only one disjunction,
+			// then the access is the disjunction.
+			if s.Disjunctions.Len() == 1 {
+				onlyDisjunction := s.Disjunctions.Oldest().Value
+				return EntitlementSetAccess{
+					Entitlements: onlyDisjunction,
+					SetKind:      Disjunction,
+				}
+			}
+
+			// There are no entitlements, but disjunctions.
+			// Allocate a new ordered map for all entitlements in the disjunctions
+			// (at minimum there are two entitlements in each disjunction).
+			entitlements = orderedmap.New[EntitlementOrderedSet](s.Disjunctions.Len() * 2)
+		}
+
+		// Add all entitlements in the disjunctions to the entitlements
+		s.Disjunctions.Foreach(func(_ string, disjunction *EntitlementOrderedSet) {
+			entitlements.SetAll(disjunction)
+		})
+	}
+
+	if entitlements == nil {
+		return UnauthorizedAccess
+	}
+
+	return EntitlementSetAccess{
+		Entitlements: entitlements,
+		SetKind:      Conjunction,
```

Comment:
To check: the behavior of this new set should be exactly the same as the old behavior except in the case where there are no conjunctions and only one disjunction? I.e. some composite type with functions `foo` and `bar` with `access(A)` and `access(B | C)` would still return `(A, B, C)` for its supported entitlements?

If we are committed to this behavior IMO we should provide some kind of warning to users when this happens, so that they are aware that their staged contract may be vulnerable to privilege escalation as a result of this behavior.

Reply:
> the behavior of this new set should be exactly the same as the old behavior except in the case where there are no conjunctions and only one disjunction?

Correct. It's not a big improvement, but at least improves things a bit. This particular case/set can actually be the result of multiple different situations in the program, e.g. `access(A | B)` occurring on multiple members of a composite, or interface types of an intersection.

> some composite type with functions `foo` and `bar` with `access(A)` and `access(B | C)` would still return `(A, B, C)` for its supported entitlements?

Correct, there is nothing we can do in such a case, because we cannot represent it in an access/auth.
IDK if it's safe and if we want to minimize disjunctions further, e.g. `(A | B), (B | C)` into `B`.

> If we are committed to this behavior IMO we should provide some kind of warning to users when this happens, so that they are aware that their staged contract may be vulnerable to privilege escalation as a result of this behavior.

Yes, we should do this in any case, with this PR or not: Developers should check that the entitlement migration does not grant more than needed.